### DISCUSSION TITLE
Can't store MigrationStrategy, which uses GenerateColumn incompatible with current Schema

### DISCUSSION THREAD

**Discussion by givip** [simolus3/drift#1620] [Q&A] [answered]
Schema v1
```
class Todo extends Table {
  IntColumn get id => integer()();

  @override
  Set<Column>? get primaryKey => {id};
}
```

Schema v2
```
class Todo extends Table {
  IntColumn get id => integer()();
  TextColumn get newCol => text().nullable()();

  @override
  Set<Column>? get primaryKey => {id};
}
```

OnUpgrade from v1 to v2 
```
await m.addColumn(todo, todo.newCol);
```

then, I rename this new column in next schema

Schema v3
```
class Todo extends Table {
  IntColumn get id => integer()();
  TextColumn get newColRenamed => text().nullable()();

  @override
  Set<Column>? get primaryKey => {id};
}
```

OnUpgrade from v2 to v3
```
m.renameColumn(todo, 'new_col', todo.newColRenamed);
```

---------

### Issue

I can't store all variation of migration code simultaneously, because  `GeneratedColumn` are different for different schemes, i.e. `OnUpgrade_1_2` uses `newCol` generated column, but after `OnUpgrade_2_3` it doesn't exist.

I should be able to upgrade from any old version to the latest.

I planned to use sequence of migration, i.e. to update schema from 1 to 4, I'd use migrations (1 -> 2), (2 -> 3), (3 -> 4), one after another?

Is this right approach?

**Comment by kuhnroyal** [comment, marked as answer]
That's not possible when using generated classes for the migration. You only have the classes for the latest schema. I would suggest using plain SQL for the updates.

**Reply by simolus3** [reply]
I'd probably go with plain SQL too.

You can use [schema exports](https://drift.simonbinder.eu/docs/advanced-features/migrations/#verifying-migrations) to generate Dart classes for old versions of your schema too. In that case you would have the older schemata around, but it takes more to set up.

**Reply by givip** [reply]
I already tried, it works, but increases binary size, I wouldn't go this way. But technically it works.

**Reply by kuhnroyal** [reply]
Yea that is why in more complex apps I would always argue to go with drift files which is basically a superset of SQL, this allows you to just copy most of the SQL for migrations.
